pkg/k8s/client/restConfig_provider.go

```diff
@@ -0,0 +1,425 @@
+		}
+	}
+
+	return &manager, err
+}
+
+// createConfig creates a rest.Config for connecting to k8s api-server.
+//
+// The precedence of the configuration selection is the following:
+// 1. kubeCfgPath
+// 2. apiServerURL(s) (https if specified)
+// 3. rest.InClusterConfig().
+func (r *restConfigManager) createConfig(cfg Config, userAgent string) (*rest.Config, error) {
+	var (
+		config       *rest.Config
+		err          error
+		apiServerURL string
+	)
+	if cfg.K8sAPIServer != "" {
+		apiServerURL = cfg.K8sAPIServer
+	} else if len(r.apiServerURLs) > 0 {
+		apiServerURL = r.apiServerURLs[0].String()
+	}
+	kubeCfgPath := cfg.K8sKubeConfigPath
+	qps := cfg.K8sClientQPS
+	burst := cfg.K8sClientBurst
+
+	switch {
+	// If the apiServerURL and the kubeCfgPath are empty then we can try getting
+	// the rest.Config from the InClusterConfig
+	case apiServerURL == "" && kubeCfgPath == "":
+		if config, err = rest.InClusterConfig(); err != nil {
+			return nil, err
+		}
+	case kubeCfgPath != "":
+		if config, err = clientcmd.BuildConfigFromFlags("", kubeCfgPath); err != nil {
+			return nil, err
+		}
+	case strings.HasPrefix(apiServerURL, "https://"):
+		if config, err = rest.InClusterConfig(); err != nil {
+			return nil, err
+		}
+		config.Host = apiServerURL
+	default:
+		//exhaustruct:ignore
+		config = &rest.Config{Host: apiServerURL, UserAgent: userAgent}
+	}
+
+	// The HTTP round tripper rotates API server URLs in case of connectivity failures.
+	if len(r.apiServerURLs) > 1 {
+		config.Wrap(r.WrapRoundTripper)
+	}
+
+	setConfig(config, userAgent, qps, burst)
+	return config, nil
+}
+
+func (r *restConfigManager) parseConfig(cfg Config) {
+	if cfg.K8sAPIServer != "" {
+		var (
+			serverURL *url.URL
+			err       error
+		)
+		s := cfg.K8sAPIServer
+		if !strings.HasPrefix(s, "http") {
+			s = fmt.Sprintf("http://%s", s) // default to HTTP
+		}
+		serverURL, err = url.Parse(s)
+		if err != nil {
+			r.log.WithError(err).Errorf("Failed to parse APIServerURL %s, skipping", serverURL)
+			return
+		}
+		r.apiServerURLs = append(r.apiServerURLs, serverURL)
+		return
+	}
+	for _, apiServerURL := range cfg.K8sAPIServerURLs {
+		if apiServerURL == "" {
+			continue
+		}
+
+		if !strings.HasPrefix(apiServerURL, "http") && !strings.HasPrefix(apiServerURL, "https") {
+			apiServerURL = fmt.Sprintf("https://%s", apiServerURL)
+		}
+
+		serverURL, err := url.Parse(apiServerURL)
+		if err != nil {
+			r.log.WithError(err).Errorf("Failed to parse APIServerURL %s, skipping", apiServerURL)
+			continue
+		}
+
+		r.apiServerURLs = append(r.apiServerURLs, serverURL)
+	}
+}
+
+func setConfig(config *rest.Config, userAgent string, qps float32, burst int) {
+	if userAgent != "" {
+		config.UserAgent = userAgent
+	}
+	if qps != 0.0 {
+		config.QPS = qps
+	}
+	if burst != 0 {
+		config.Burst = burst
+	}
+}
+
+func (r *restConfigManager) rotateAPIServerURL() {
+	if len(r.apiServerURLs) <= 1 {
+		return
+	}
+
+	r.rt.Lock()
```

Comment:
You have nested locks here; is there any chance of a deadlock? Is there anything else that takes `r.rt.Mutex` and `r.Mutex`?

Reply:
No, as long as the locks are taken in the same order.